### DISCUSSION TITLE
Modular pages

### DISCUSSION THREAD

**Discussion by taviroquai** [vercel/next.js#21755] [Ideas]
Hi,

Considering that Next pages are static files, and all files inside pages folder are routed as pages (which may not be desired), I'm trying to build a modular architecture in a Nextjs app, or putting as simple as this: modular pages.

Let's consider a git submodules example.

**Main repo:**
./pages/index.js
./components/Layout.js

```javascript
/* index.js page */
import Layout form '../components/Layout';
export default function Home() {
  return <Layout />
}
```

**Submodule repo:**
./pages/moduleA.js
./moduleA/components/Button.js

```javascript
/* moduleA.js page */
import Button form '../moduleA/components/Button';
export default function ModuleA() {
  return <Button />
}
```

The problem with this approach, is that the submodule will live in one folder (let say pages/moduleA), and not as intertwined files and folder in the main repo, which would allow to add/removes pages as modules.

Any ideias on this? Thanks!

**Comment by m-s7** [comment]


**Reply by josh-sachs-epic** [reply]
@michalsmolik - could you elaborate on how you were able to get nextjs routing working for files defined outside of a `pages` folder?

What I'd like to be able to do is define a pattern for feature submodules within my nextjs project. 

consider
```
/src
-- /components
-- /pages
---- /api
---- index.tsx
-- /features
---- /feature_a
------ /components
------ /api
-------- get_data.tsx
------ /pages
-------- index.tsx
-------- another_page.tsx
---- /feature_b
------ /components
------ /api
------ /pages
```

this is a rough outline... but i think conveys the pattern well.   If nextjs offered any real flexibility in this regard, each feature could export a known interface via `index.ts` like: 

```
import get_data from "./api/get_data"
export default feature = {
  name: "feature_a",
  root: "/feature_a",
  routes: [
    { type: "api", path: "/get_data", handler: get_data },
    ...   
  ],
}
```

Anyway, since that doesn't seem possible - I'm curious to see other creative solutions to adding modularity to nextjs project structure. 

Looking forward to any insights you've got!

**Comment by yalegria** [comment]
I have a similar need. I created a subfolder inside the **pages** folder where I add my git submodules. The big question is what is the next step? How can I integrate the submodule functionality within the parent application? Just putting the right routing does not work, it gives a 404.
